### PR TITLE
Save Site Profiler state to redux to persist state when changing routes 

### PR DESCRIPTION
**Pull request by candy02058912**
Fixes https://github.com/Automattic/dotcom-forge/issues/9244

## Proposed Changes

* Fixed retest button
* Save hash to redux
* Save page state to redux
* Adjusted the steps so that it makes sense

https://github.com/user-attachments/assets/bbdc5c46-d46c-41b2-b093-41a9028184e8

Note: The progress steps were all "fake," probably for UX reasons, as the API endpoint only has three statuses: `queued`, `running`, and `completed`.

## Why are these changes being made?

* Better user experience

## Testing Instructions

* Go to `calypso.localhost:3000/sites/performance/:site`
* Run a test, wait for the first step to complete
* Switch to another tab
* Switch back to the tab and the test should not let you feel like it's starting over
* Check nothing is broken at `calypso.localhost:3000/speed-test-tool?url=<site>`

Remember that this is a "perceived" progress, so acceptance criteria is as long as the user experience "feels smooth".

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Review by paulopmt1 — COMMENTED** [review]
Shared a review here: p1728308125141749-slack-C04H4NY6STW

**Comment by candy02058912** [comment]
Would like to note that when we save the state of the pages in the `lastVisitedReport`, we're not reflecting that in the query params. Unless there is an easy way to do that, I don't think we'll be exploring that due to alignment on this work has been in a "diminishing return" state.

**Comment by p-jackson** [comment]
> Would like to note that when we save the state of the pages in the `lastVisitedReport`, we're not reflecting that in the query params. Unless there is an easy way to do that, I don't think we'll be exploring that due to alignment on this work has been in a "diminishing return" state.

**Comment by p-jackson** [comment]
woops!

**Comment by autumnfjeld** [comment]
Looking through the copy changes:


* These commits added/changed the copy 
  *   [Break the steps logic into logged in and logged out](https://github.com/Automattic/wp-calypso/pull/95151/commits/1cdc37a9038ffeab7883fcb67650bddb19ea92e8)
  * [Add a new loading state for logged in site profiler](https://github.com/Automattic/wp-calypso/pull/95151/commits/e3c57df21039425fa3540ee779f33f57a333cc83)


image

image

If we are going to change the copy then let's do it in a different PR with clear intention and screenshots showing before and after.

Unless these changes were tied to fixing a performance issue and that is the motivation for them?    

_____
Side note: I'd remove the ellipses.

**Comment by autumnfjeld** [comment]
Follow up, watched @candy02058912 's explainer video: p1728470760108529/1728445740.969029-slack-C04H4NY6STW
Thank you @vykes-mac for reminding me she explained there. 

I don't have strong opinions.  Except remove the elipses. 

Let's be consistent. All the steps are some sort of waiting, we don't need elipses on the end. 😄 

image

**Comment by vykes-mac** [comment]
Also ensure this new copy is only shown in the logged-in version, logged-out should behave the same.

File: client/performance-profiler/pages/loading-screen/progress.tsx [hunk 1]
```diff
@@ -115,18 +117,30 @@ const useLoadingSteps = ( {
 	if ( isLoadingPages ) {
 		steps = [ translate( 'Getting your site pages' ) ];
 	} else {
-		steps = isSavedReport
+		const loggedInSteps = [
+			...( step === 0
+				? [ translate( 'Checking for existing report…' ) ]
```
Comment (matt-west) [inline client/performance-profiler/pages/loading-screen/progress.tsx:122]: I don’t think we use "report" anywhere else in the copy, so let’s tweak these messages to:

> Checking for recent performance tests

> No recent tests found

**Comment by matt-west** [comment]
Thanks @candy02058912! I appreciate the attention to detail here.

I think it makes sense to add the extra step when checking for existing results. I left a [comment](https://github.com/Automattic/wp-calypso/pull/95151#discussion_r1794990661) about tweaking the copy.

Ideally we would only show the rest of the loading steps once we’ve determined that existing results are not available. It feels a little weird to me to show everything right away, as we don’t yet know that those other steps are needed. Here’s what that would look like:

| Checking for existing results | Running a full test |
|--------|--------|
| image | image | 

If that’s going to overcomplicate the implementation, we can just show all the steps at once.

Happy to drop the ellipsis as @autumnfjeld suggested.

---

**Additional context**

The original version of the logged-out tool had a standard loading screen, but we received some feedback in early user tests that it felt the results took a long time to load. The fake loading steps were introduced to give the perception of progress so users didn’t feel like the test was stuck.

File: client/state/site-profiler/reducer.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+} from 'calypso/state/action-types';
+
+const reportReducer = ( state = {}, action ) => {
+	const { siteId, url, hash, pageId } = action;
+	switch ( action.type ) {
+		case SITE_PROFILER_SET_REPORT:
+			return {
+				...state,
+				[ siteId ]: {
+					...( state[ siteId ] || {} ),
```
Comment (p-jackson) [inline client/state/site-profiler/reducer.js:14]: JS fact: you can actually use the spread operator on `undefined` or `null` values and they do the same thing as spreading an empty object 👍 

```suggestion
					...state[ siteId ],
```

**Comment by candy02058912** [comment]
We've decided to move forward with the implementation here: https://github.com/Automattic/wp-calypso/pull/95245

**Pull request closed without merging**